src/metabase/api/collection.clj

```diff
@@ -187,18 +207,40 @@
   [id]
   (collection-detail (api/read-check Collection id)))
 
+(defn- coerce-vec [to-coerce]
+  (if (and (some? to-coerce) (not (vector? to-coerce))) [to-coerce] to-coerce))
+
 (api/defendpoint GET "/:id/items"
   "Fetch a specific Collection's items with the following options:
 
-  *  `model` - only include objects of a specific `model`. If unspecified, returns objects of all models
-  *  `archived` - when `true`, return archived objects *instead* of unarchived ones. Defaults to `false`."
-  [id model archived]
-  {model    (s/maybe (apply s/enum valid-model-param-values))
-   archived (s/maybe su/BooleanString)}
-  (collection-children
-    (api/read-check Collection id)
-    {:model     (keyword model)
-     :archived? (Boolean/parseBoolean archived)}))
+  *  `models` - only include objects of a specific set of `models`. If unspecified, returns objects of all models
+  *  `archived` - when `true`, return archived objects *instead* of unarchived ones. Defaults to `false`.
+  *  `pinned_state` - when `is_pinned`, return pinned objects only.
+                   when `is_not_pinned`, return non pinned objects only.
+                   when `all`, return everything.
+  *  `limit` - limit for pagination.
+  *  `offset` - offset for pagination."
+  [id models archived pinned_state limit offset]
+  {models       (s/maybe models-schema)
+   archived     (s/maybe su/BooleanString)
+   pinned_state (s/maybe (apply s/enum valid-pinned-state-values))
+   limit        (s/maybe su/IntStringGreaterThanZero)
+   offset       (s/maybe su/IntStringGreaterThanOrEqualToZero)}
+  (api/check-valid-page-params limit offset)
+  (let [model-kwds    (apply hash-set (map keyword (coerce-vec models)))
+        children-res  (collection-children (api/read-check Collection id)
+                                           {:models       model-kwds
+                                            :archived?    (Boolean/parseBoolean archived)
+                                            :pinned-state (keyword pinned_state)})
+        limit-int     (some-> limit Integer/parseInt)
+        offset-int    (some-> offset Integer/parseInt) ]
+    {:data   (cond->> (vec children-res)
+               (some? offset-int) (drop offset-int)
+               (some? limit-int)  (take limit-int))
```

Comment:
It seems a little inefficient to fetch everything and then drop them in Clojure land rather than doing that at in the DB query itself... any reason we can't do it that way? You would have to rework `collection-children` a bit to do it.

At the very least, if `limit` is `10` and you specify multiple models (let's say `dashboard` and `card`, and we fetch 10 `dashboards` in the first query, we shouldn't be running another query to fetch Cards if we're going to throw those all out. Or if `limit` is `10` and we get `5` Dashboards, the subsequent call to fetch `Cards` should have `LIMIT 5`.

Reply:
I did rewrite `collection-children` to do it initially and then it got out of hand with the incorporation of the left join queries. I could try taking stuff lazily from reducible queries

Reply:
I think this is performance-impinging enough to actually do, now that I think about it. Willfix